### PR TITLE
Port si7210 driver with new driver framework

### PR DESCRIPTION
**Pull request by Jaeyoung-Lim**
**Describe problem solved by this pull request**
This commit ports the si7210 hall sensor to the new driver framework, since https://github.com/ethz-asl/ethzasl_fw_px4/pull/17 is unable to build with the latest firmware.

This is implemented on top of https://github.com/ethz-asl/ethzasl_fw_px4/pull/17

This was done by using the SDP3X airspeed sensor as reference. Although this is giving out measurements from the sensor, I am not sure if I have removed any critical functionalities from the old driver.

@ASM3 It would be very helpful if you can point out some of the things that I might have done wrong :smile: 

**Test data / coverage**
Tested on Pixhawk4 + vane sensor
```
TOPIC: sensor_hall
 sensor_hall_s
        timestamp: 211557475  (0.004923 seconds ago)
        mag_t: -16.4975
        temp_c: 26.2691
        instance: 0
nsh> listener sensor_hall

TOPIC: sensor_hall
 sensor_hall_s
        timestamp: 212413644  (0.014693 seconds ago)
        mag_t: -10.5500
        temp_c: 26.5596
        instance: 0
nsh> listener sensor_hall

TOPIC: sensor_hall
 sensor_hall_s
        timestamp: 213460567  (0.007871 seconds ago)
        mag_t: -4.2100
        temp_c: 26.5596
        instance: 0
```

**Additional context**
- This PR is against https://github.com/ethz-asl/ethzasl_fw_px4/pull/17
- The probing doesn't include searching for multiple I2C addresses. Probably this is the reason why it is only working with `I2C_ADDRESS_2_SI7210`

### CODE REVIEW

**Review by ASM3 — CHANGES_REQUESTED** [review]
General remark: functionality, boot sequence and timing could not be verified as the HW is not available

File: src/drivers/si7210/si7210.cpp [hunk 1]
```diff
@@ -40,545 +40,174 @@
...
-		PX4_ERR("driver poll restart failed");
-		exit(1);
+	if (require_initialization && _keep_retrying) {
+		PX4_INFO("no sensor found, but will keep retrying");
+		return 0;
 	}
 
-	exit(0);
+	return require_initialization ? -1 : 0;
 }
```
[the remark this answers is not in the document]
Comment (acfloria) [reply src/drivers/si7210/si7210.cpp:56]: I checked it with some print statements. This probe is called but not the base Vane::probe one.
I also found that probably (due to this) the starting without keep retrying is not very reliable.

File: src/drivers/si7210/CMakeLists.txt [hunk 2]
```diff
@@ -34,9 +34,12 @@ px4_add_module(
 	MODULE drivers__si7210
 	MAIN si7210
 	COMPILE_FLAGS
-			-O0
+		# -DDEBUG_BUILD
```
[the remark this answers is not in the document]
Comment (Jaeyoung-Lim) [reply src/drivers/si7210/CMakeLists.txt]: Done!

File: src/drivers/si7210/si7210.cpp [hunk 3]
```diff
@@ -40,545 +40,174 @@
...
-int
-SI7210::collect()
+int SI7210::collect()
 {
+	perf_begin(_sample_perf);
+
 	_collect_phase = false;
-	bool si7210_notify = true;
-	si7210_report  prb;
+	// bool si7210_notify = true;
```
[the remark this answers is not in the document]
Comment (Jaeyoung-Lim) [reply src/drivers/si7210/si7210.cpp]: Removed comment

File: src/drivers/si7210/si7210.h [hunk 4]
```diff
@@ -276,9 +175,19 @@ class SI7210 : public device::I2C
...
+	bool crc(const uint8_t data[], unsigned size, uint8_t checksum);
+
+	/**
+	 * Write a command in Sensirion specific logic
+	 */
+	int write_command(uint16_t command);
+
+	uint16_t _scale{0};
+	const bool _keep_retrying;
+	State _state{State::RequireConfig};
```
[the remark this answers is not in the document]
Comment (Jaeyoung-Lim) [reply src/drivers/si7210/si7210.h]: Actually not, I have removed the crc now, thanks!

File: src/drivers/si7210/si7210.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ASM3) [reply src/drivers/si7210/si7210.cpp]: great

**Comment by ASM3** [comment]
lgtm

**Comment by Jaeyoung-Lim** [comment]
@ASM3 Thanks! Merging!

**Pull request merged**